momaland/utils/aec_wrappers.py

```diff
@@ -0,0 +1,87 @@
+"""Various wrappers for AEC MO environments."""
+
+import numpy as np
+from gymnasium.wrappers.normalize import RunningMeanStd
+from pettingzoo.utils.wrappers.base import BaseWrapper
+
+
+class LinearizeReward(BaseWrapper):
+    """Convert MO reward vector into scalar SO reward value.
+
+    `weights` represents the weights of each objective in the reward vector space.
+    """
+
+    def __init__(self, env, weights: dict):
+        """Reward linearization class initializer.
+
+        Args:
+            env: base env to add the wrapper on.
+            weights: a dict where keys are agents and values are vectors representing the weights of their rewards.
+        """
+        self.weights = weights
+        super().__init__(env)
+
+    def last(self, observe: bool = True):
+        """Returns a reward scalar from the reward vector."""
+        observation, rewards, termination, truncation, info = self.env.last(observe=observe)
+        if self.env.agent_selection in list(self.weights.keys()):
+            rewards = np.array([np.dot(rewards, self.weights[self.env.agent_selection])])
+        return observation, rewards, termination, truncation, info
+
+
+class NormalizeReward(BaseWrapper):
+    r"""This wrapper will normalize immediate rewards s.t. their exponential moving average has a fixed variance.
+
+    The exponential moving average will have variance :math:`(1 - \gamma)^2`.
+
+    Note:
+        The scaling depends on past trajectories and rewards will not be scaled correctly if the wrapper was newly
+        instantiated or the policy was changed recently.
+    """
+
+    def __init__(
+        self,
+        env,
+        agent,
+        idx,
+        gamma: float = 0.99,
+        epsilon: float = 1e-8,
+    ):
+        """This wrapper will normalize immediate rewards s.t. their exponential moving average has a fixed variance.
+
+        Args:
+            env: The environment to apply the wrapper
+            agent: the agent whose reward will be normalized
+            idx: the index of the rewards that will be normalized.
+            epsilon: A stability parameter
+            gamma: The discount factor that is used in the exponential moving average.
+        """
+        super().__init__(env)
+        self.agent = agent
+        self.idx = idx
+        self.return_rms = RunningMeanStd(shape=())
+        self.returns = np.array([0.0])
```

Comment:
Not sure to understand why this is an array and not a scalar?

Reply:
Gymnasium RMS doesnt like it when i give a scalar. works when it's a single element array. I could try to figure it why but it works like this...